File: go/applieddiscount.go
// ABOUTME: AppliedDiscount domain type representing a discount applied to a cart or line item.
// ABOUTME: Preserves explainability with promotion ID, amount, target, details, and optional allocations.
package promotionengine

// AppliedDiscount represents a discount that has been applied.
// It preserves explainability by tracking the promotion, amount, target, and details.
type AppliedDiscount struct {
	PromotionID PromotionID
	Amount      Money
	Target      string // "line" or "cart"
	Details     string
	Allocations map[Sku]Money // Optional: per-SKU discount allocations
}

// NewAppliedDiscount creates a new AppliedDiscount without allocations.
func NewAppliedDiscount(promotionID PromotionID, amount Money, target, details string) AppliedDiscount {
	return AppliedDiscount{
		PromotionID: promotionID,
		Amount:      amount,
		Target:      target,
		Details:     details,
		Allocations: nil,
	}
}

// NewAppliedDiscountWithAllocations creates a new AppliedDiscount with per-SKU allocations.
func NewAppliedDiscountWithAllocations(promotionID PromotionID, amount Money, target, details string, allocations map[Sku]Money) AppliedDiscount {
	return AppliedDiscount{
		PromotionID: promotionID,
		Amount:      amount,
		Target:      target,
		Details:     details,
		Allocations: allocations,
	}
}

File: java/src/main/java/com/promoengine/pricing/AppliedDiscount.java
// ABOUTME: Represents a discount that has been applied to a cart with full explainability.
// ABOUTME: Captures promotion ID, amount, target scope, details, and optional per-SKU allocations.
package com.promoengine.pricing;

import com.promoengine.domain.Money;
import com.promoengine.domain.PromotionId;
import com.promoengine.domain.Sku;

import java.util.Map;

public record AppliedDiscount(
    PromotionId promotionId,
    Money amount,
    String target,
    String details,
    Map<Sku, Money> allocations
) {

    public AppliedDiscount {
        if (promotionId == null) {
            throw new IllegalArgumentException("Promotion ID cannot be null");
        }
        if (amount == null) {
            throw new IllegalArgumentException("Amount cannot be null");
        }
        if (target == null || target.trim().isEmpty()) {
            throw new IllegalArgumentException("Target cannot be null or empty");
        }
        if (details == null) {
            throw new IllegalArgumentException("Details cannot be null");
        }
        if (allocations == null) {
            allocations = Map.of();
        } else {
            // Defensive copy to ensure immutability
            allocations = Map.copyOf(allocations);
        }
    }

    // Convenience constructor without allocations
    public AppliedDiscount(PromotionId promotionId, Money amount, String target, String details) {
        this(promotionId, amount, target, details, Map.of());
    }

    // TODO: Consider adding "reason" field for why promotion was not applied
    // TODO: Consider adding "evaluationTrace" for detailed decision logging
}
